cpp/src/arrow/tensor.h

```diff
@@ -114,6 +114,34 @@ class ARROW_EXPORT Tensor {
   ARROW_DISALLOW_COPY_AND_ASSIGN(Tensor);
 };
 
+template <typename TYPE>
+class ARROW_EXPORT NumericTensor : public Tensor {
+ public:
+  using TypeClass = TYPE;
+  using value_type = typename TypeClass::c_type;
+
+  /// Constructor with no dimension names or strides, data assumed to be row-major
+  NumericTensor(const std::shared_ptr<Buffer>& data, const std::vector<int64_t>& shape);
+
+  /// Constructor with non-negative strides
+  NumericTensor(const std::shared_ptr<Buffer>& data, const std::vector<int64_t>& shape,
+                const std::vector<int64_t>& strides);
+
+  /// Constructor with strides and dimension names
+  NumericTensor(const std::shared_ptr<Buffer>& data, const std::vector<int64_t>& shape,
+                const std::vector<int64_t>& strides,
+                const std::vector<std::string>& dim_names);
+
+  const value_type& Value(const std::vector<int64_t>& index) const {
+    int64_t offset = CalculateValueOffset(index);
+    const value_type* ptr = reinterpret_cast<const value_type*>(raw_data() + offset);
+    return *ptr;
```

Comment:
On some CPU architectures, this assumes the data is naturally aligned. It probably doesn't matter for now.

Reply:
One way to tackle this is to make CalculateValueOffset return an offset in units of TYPE instead of bytes. It'll also make it easier to work with non-fixed-bytes in the future (also possibly correcting the problem with BOOL).

Reply:
I changed this function to return an offset in units of TYPE.

Reply:
I see.